File: sdk/npm/src/providers/tmpmails.ts
import { InternalEmailInfo, Email, Channel } from '../types';
import { normalizeEmail } from '../normalize';
import { fetchWithTimeout } from '../retry';

const CHANNEL: Channel = 'tmpmails';
const ORIGIN = 'https://tmpmails.com';
const TOKEN_SEP = '\t';

const EMAIL_RE = /[a-zA-Z0-9._-]+@tmpmails\.com/g;
const PAGE_CHUNK_RE = /\/_next\/static\/chunks\/app\/%5Blocale%5D\/page-[a-f0-9]+\.js/;
const INBOX_ACTION_RE = /"([0-9a-f]+)",[^,]+,void 0,[^,]+,"getInboxList"/;

function localeFromDomain(domain?: string | null): string {
  const s = String(domain ?? '').trim();
  return s || 'zh';
}

function setCookieLines(headers: Headers): string[] {
  const h = headers as Headers & { getSetCookie?: () => string[] };
  if (typeof h.getSetCookie === 'function') {
    return h.getSetCookie();
  }
  const one = headers.get('set-cookie');
  return one ? [one] : [];
}

function parseUserSign(headers: Headers): string {
  for (const line of setCookieLines(headers)) {
    const m = /^user_sign=([^;]+)/i.exec(line);
    if (m) {
      const v = m[1].trim();
      if (v) return v;
    }
  }
  throw new Error('tmpmails: missing user_sign cookie');
}

function pickEmail(html: string): string {
  const support = 'support@tmpmails.com';
  const counts = new Map<string, number>();
  for (const m of html.matchAll(EMAIL_RE)) {
    const addr = m[0];
    if (addr.toLowerCase() === support) continue;
    counts.set(addr, (counts.get(addr) ?? 0) + 1);
  }
  if (counts.size === 0) {
    throw new Error('tmpmails: no inbox address in page');
  }
  let best = '';
  let bestC = 0;
  for (const [addr, c] of counts) {
    if (c > bestC || (c === bestC && addr < best)) {
      best = addr;
      bestC = c;
    }
  }
  return best;
}

async function fetchInboxActionID(html: string): Promise<string> {
  const sub = html.match(PAGE_CHUNK_RE)?.[0];
  if (!sub) throw new Error('tmpmails: page chunk script not found');
  const res = await fetchWithTimeout(`${ORIGIN}${sub}`, {
    headers: {
      'User-Agent':
        'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36 Edg/146.0.0.0',
      Accept: '*/*',
      Referer: `${ORIGIN}/`,
    },
  });
  if (!res.ok) {
    throw new Error(`tmpmails page chunk: ${res.status}`);
  }
  const js = await res.text();
  const m = js.match(INBOX_ACTION_RE);
  if (!m?.[1]) throw new Error('tmpmails: getInboxList action not found in chunk');
  return m[1];
}

export async function generateEmail(domain?: string | null): Promise<InternalEmailInfo> {
  const loc = localeFromDomain(domain);
  const pageUrl = `${ORIGIN}/${loc}`;
  const res = await fetchWithTimeout(pageUrl, {
    headers: {
      'User-Agent':
        'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36 Edg/146.0.0.0',
      Accept: 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
      'Accept-Language': 'zh-CN,zh;q=0.9,en;q=0.8,en-GB;q=0.7,en-US;q=0.6',
      'Cache-Control': 'no-cache',
      DNT: '1',
      Pragma: 'no-cache',
      Referer: pageUrl,
      'Upgrade-Insecure-Requests': '1',
    },
  });
  if (!res.ok) {
    throw new Error(`tmpmails generate: ${res.status}`);
  }
  const html = await res.text();
  const userSign = parseUserSign(res.headers);
  const email = pickEmail(html);
  const actionId = await fetchInboxActionID(html);
  const token = `${loc}${TOKEN_SEP}${userSign}${TOKEN_SEP}${actionId}`;
  return { channel: CHANNEL, email, token };
}

function parseInboxResponse(recipient: string, raw: string): Email[] {
  const text = raw.split('"$undefined"').join('null');
  const out: Email[] = [];
  let dataErr: Error | null = null;
  for (const line of text.split('\n')) {
    const trimmed = line.trim();
    if (!trimmed) continue;
    const colon = trimmed.indexOf(':');
    if (colon <= 0) continue;
    const jsonPart = trimmed.slice(colon + 1).trim();
    if (!jsonPart.startsWith('{')) continue;
    let wrap: { code?: number; data?: { list?: unknown[] } };
    try {
      wrap = JSON.parse(jsonPart) as { code?: number; data?: { list?: unknown[] } };
    } catch {
      continue;
    }
    if (wrap.code !== 200 || !wrap.data) continue;
    const list = wrap.data.list;
    if (!Array.isArray(list)) {
      dataErr = new Error('tmpmails: invalid list');
      continue;
    }
    for (const item of list) {
      if (item && typeof item === 'object') {
        out.push(normalizeEmail(item, recipient));
      }
    }
    return out;
  }
  if (dataErr) throw dataErr;
  throw new Error('tmpmails: inbox payload not found');
}

export async function getEmails(email: string, token: string): Promise<Email[]> {
  const parts = token.split(TOKEN_SEP);
  if (parts.length !== 3) {
    throw new Error('tmpmails: invalid session token');
  }
  const [locale, userSign, actionId] = parts;
  const postUrl = `${ORIGIN}/${locale}`;
  const body = JSON.stringify([userSign, email, 0]);
  const res = await fetchWithTimeout(postUrl, {
    method: 'POST',
    headers: {
      'User-Agent':
        'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36 Edg/146.0.0.0',
      Accept: 'text/x-component',
      'Content-Type': 'text/plain;charset=UTF-8',
      'Next-Action': actionId,
      Origin: ORIGIN,
      Referer: postUrl,
      Cookie: `NEXT_LOCALE=${locale}; user_sign=${userSign}`,
    },
    body,
  });
  if (!res.ok) {
    throw new Error(`tmpmails inbox: ${res.status}`);
  }
  const raw = await res.text();
  return parseInboxResponse(email, raw);
}

File: sdk/python/tempmail_sdk/providers/tmpmails.py
"""
tmpmails.com：GET 首页下发 user_sign Cookie 与页面中的临时地址；
收信为 Next.js Server Action POST（Accept: text/x-component），body 为 [user_sign, email, 0]。
"""

import json
import re
from typing import List, Optional

from .. import http as tm_http
from ..normalize import normalize_email
from ..types import EmailInfo, Email

CHANNEL = "tmpmails"
ORIGIN = "https://tmpmails.com"
TOKEN_SEP = "\t"

_EMAIL_RE = re.compile(r"[a-zA-Z0-9._-]+@tmpmails\.com")
_PAGE_CHUNK_RE = re.compile(r"/_next/static/chunks/app/%5Blocale%5D/page-[a-f0-9]+\.js")
_INBOX_ACTION_RE = re.compile(
    r'"([0-9a-f]+)",[^,]+,void 0,[^,]+,"getInboxList"'
)

_DEFAULT_HEADERS = {
    "User-Agent": (
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) "
        "Chrome/146.0.0.0 Safari/537.36 Edg/146.0.0.0"
    ),
    "Accept-Language": "zh-CN,zh;q=0.9,en;q=0.8,en-GB;q=0.7,en-US;q=0.6",
    "Cache-Control": "no-cache",
    "DNT": "1",
    "Pragma": "no-cache",
}


def _locale(domain: Optional[str]) -> str:
    s = (domain or "").strip()
    return s or "zh"


def _user_sign_from_response(resp) -> str:
    for c in resp.cookies.items():
        if str(c[0]).lower() == "user_sign":
            v = str(c[1]).strip()
            if v:
                return v
    raise RuntimeError("tmpmails: missing user_sign cookie")


def _pick_email(html: str) -> str:
    support = "support@tmpmails.com"
    counts: dict[str, int] = {}
    for m in _EMAIL_RE.finditer(html):
        addr = m.group(0)
        if addr.lower() == support:
            continue
        counts[addr] = counts.get(addr, 0) + 1
    if not counts:
        raise RuntimeError("tmpmails: no inbox address in page")
    return sorted(counts.items(), key=lambda kv: (-kv[1], kv[0]))[0][0]


def _fetch_inbox_action_id(html: str) -> str:
    m = _PAGE_CHUNK_RE.search(html)
    if not m:
        raise RuntimeError("tmpmails: page chunk script not found")
    sub = m.group(0)
    url = ORIGIN + sub
    r = tm_http.get(
        url,
        headers={
            **_DEFAULT_HEADERS,
            "Accept": "*/*",
            "Referer": ORIGIN + "/",
        },
    )
    r.raise_for_status()
    m2 = _INBOX_ACTION_RE.search(r.text)
    if not m2:
        raise RuntimeError("tmpmails: getInboxList action not found in chunk")
    return m2.group(1)


def generate_email(domain: Optional[str] = None, **kwargs) -> EmailInfo:
    loc = _locale(domain)
    page_url = f"{ORIGIN}/{loc}"
    resp = tm_http.get(
        page_url,
        headers={
            **_DEFAULT_HEADERS,
            "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
            "Referer": page_url,
            "Upgrade-Insecure-Requests": "1",
        },
    )
    resp.raise_for_status()
    html = resp.text
    user_sign = _user_sign_from_response(resp)
    email = _pick_email(html)
    action_id = _fetch_inbox_action_id(html)
    token = TOKEN_SEP.join((loc, user_sign, action_id))
    return EmailInfo(channel=CHANNEL, email=email, _token=token)


def _parse_inbox_response(recipient: str, raw: str) -> List[Email]:
    text = raw.replace('"$undefined"', "null")
    data_err = None
    for line in text.split("\n"):
        line = line.strip()
        if not line or ":" not in line:
            continue
        _, rest = line.split(":", 1)
        json_part = rest.strip()
        if not json_part.startswith("{"):
            continue
        try:
            wrap = json.loads(json_part)
        except json.JSONDecodeError:
            continue
        if wrap.get("code") != 200:
            continue
        data = wrap.get("data") or {}
        raw_list = data.get("list")
        if not isinstance(raw_list, list):
            data_err = RuntimeError("tmpmails: invalid list")
            continue
        out = []
        for item in raw_list:
            if isinstance(item, dict):
                out.append(normalize_email(item, recipient))
        return out
    if data_err:
        raise data_err
    raise RuntimeError("tmpmails: inbox payload not found")


def get_emails(email: str, token: str, **kwargs) -> List[Email]:
    parts = token.split(TOKEN_SEP)
    if len(parts) != 3:
        raise ValueError("tmpmails: invalid session token")
    locale, user_sign, action_id = parts
    post_url = f"{ORIGIN}/{locale}"
    body = json.dumps([user_sign, email, 0], separators=(",", ":"))
    resp = tm_http.post(
        post_url,
        headers={
            **_DEFAULT_HEADERS,
            "Accept": "text/x-component",
            "Content-Type": "text/plain;charset=UTF-8",
            "Next-Action": action_id,
            "Origin": ORIGIN,
            "Referer": post_url,
            "Cookie": f"NEXT_LOCALE={locale}; user_sign={user_sign}",
        },
        data=body.encode("utf-8"),
    )
    resp.raise_for_status()
    return _parse_inbox_response(email, resp.text)
